src/main/scala/org/spartanz/parserz/InOut.scala

```diff
@@ -0,0 +1,143 @@
+      final def extractCount(s: CountAndState): Int = (s >> 32).toInt
+      final def extractState(s: CountAndState): Int = s.toInt
+    }
+  }
+
+  sealed trait Producer[+E, -A, O]
+  object Producer {
+    sealed trait Mutable[+E, -A, O] extends Producer[E, A, O] {
+      type State
+      def create(a: A): State
+      def copy(s: State): State
+      def needsMoreOutput(s: State): Boolean
+      def expel(s: State, output: O, i: Int): Int
+    }
+
+    sealed trait Simple[+E, -A, O] extends Producer[E, A, O] {
+      def create(a: A): Int
+      def needsMoreOutput(s: Int): Boolean
+      def expel(s: Int, output: O, i: Int): Long
+    }
+  }
+
+  case class One[+E, I, A](consumer: Consumer.Simple[E, I, A], condition: A => Boolean)
+  case class Many[+E, I, A](consumer: Consumer.Mutable[E, I, A])
+  case class ManyConditional[+E, I, A](consumer: Consumer.Mutable[E, I, A])
+
+
+  object Chars {
+    import Consumer._
+
+    private val single: Simple[Nothing, Array[Char], Char] = new Simple[Nothing, Array[Char], Char] {
+      final val create: Int = -1
+      final def needsMoreInput(s: Int): Boolean = s == -1
+
+      final def feed(s: Int, input: Array[Char], i: Int): Simple.CountAndState =
+        try { Simple.create(1, input(i).toInt) }
+        catch { case _: ArrayIndexOutOfBoundsException => -1L }
+
+      final def finish(s: Int): Char =
+        if (s == -1) throw NoInput
+        else s.toChar
+    }
+
+    private def multiple(p: Char => Boolean): Mutable[Nothing, Array[Char], Array[Char]] = {
+      class St(var acc: Array[Char], var done: Boolean)
+
+      new Mutable[Nothing, Array[Char], Array[Char]] {
+        final type State = St
+        final def create(): State = new St(Array.emptyCharArray, false)
+        final def copy(s: State): State = new St(s.acc, s.done)
+        final def needsMoreInput(s: State): Boolean = !s.done
+
+        final def feed(s: State, input: Array[Char], i: Int): Int = {
+          val last = input.indexWhere(!p(_), i)
+          val i2 = if (last == -1) input.length else last
+          val chunk = java.util.Arrays.copyOfRange(input, i, i2)
+
+          s.acc ++= chunk
+          s.done = last != -1 || input.length == 0
+          chunk.length
+        }
+
+        final def finish(s: State): Array[Char] =
+          s.acc
+      }
+    }
+
+    val one: One[Nothing, Array[Char], Char] =
+      One(single, null)
+
+    def oneIf(expr: Expr[Char]): One[Nothing, Array[Char], Char] =
+      One(single, Expr.exprFilter(expr))
+
+    def manyWhile[E](expr: Expr[Char]): Many[E, Array[Char], Array[Char]] =
+      Many(multiple(Expr.exprFilter(expr)))
+
+    def token[E](t: String): ManyConditional[E, Array[Char], Array[Char]] = {
+      val req = t.toCharArray
+      val len = t.length
+
+      class S(var read: Int, var res: Boolean)
+
+      ManyConditional(
+        new Mutable[E, Array[Char], Array[Char]] {
```

Comment:
Try to use simple by storing a bit into the `Int` for the boolean flag and using the rest of the bits for `read`.

Reply:
Sure, thanks for the hint!
This specific instance is neglected because it has very limited effect on my perf test result.